app/src/physical_layouts.c

```diff
@@ -0,0 +1,390 @@
+    const struct position_map_entry *dest_pos_map = NULL;
+
+    for (int pm = 0; pm < ARRAY_SIZE(positions_maps); pm++) {
+        if (positions_maps[pm].layout == src_layout) {
+            src_pos_map = &positions_maps[pm];
+        }
+
+        if (positions_maps[pm].layout == dest_layout) {
+            dest_pos_map = &positions_maps[pm];
+        }
+    }
+#endif
+
+    memset(map, UINT32_MAX, dest_layout->keys_len);
+
+    for (int b = 0; b < dest_layout->keys_len; b++) {
+        bool found = false;
+
+#if HAVE_POS_MAP
+        if (src_pos_map && dest_pos_map) {
+            for (int m = 0; m < ZMK_POS_MAP_LEN; m++) {
+                if (dest_pos_map->positions[m] == b) {
+                    map[b] = src_pos_map->positions[m];
+                    found = true;
+                    break;
+                }
+            }
+        }
+#endif
+
+#if !POS_MAP_COMPLETE
+        if (!found) {
+            const struct zmk_key_physical_attrs *key = &dest_layout->keys[b];
+            for (int old_b = 0; old_b < src_layout->keys_len; old_b++) {
+                const struct zmk_key_physical_attrs *candidate_key = &src_layout->keys[old_b];
+
+                if (candidate_key->x == key->x && candidate_key->y == key->y) {
+                    map[b] = old_b;
+                    found = true;
+                    break;
+                }
+            }
+        }
+#endif
+
+        if (!found || map[b] >= src_layout->keys_len) {
+            map[b] = UINT32_MAX;
+        }
+    }
+
+    return dest_layout->keys_len;
+}
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+
+static int physical_layouts_handle_set(const char *name, size_t len, settings_read_cb read_cb,
+                                       void *cb_arg) {
+    const char *next;
+
+    if (settings_name_steq(name, "selected", &next) && !next) {
+        if (len != sizeof(saved_selected_index)) {
+            return -EINVAL;
+        }
+
+        int err = read_cb(cb_arg, &saved_selected_index, len);
+        if (err <= 0) {
+            LOG_ERR("Failed to handle selected physical dest_layout from settings (err %d)", err);
+            return err;
+        }
+
+        return zmk_physical_layouts_select(saved_selected_index);
+    }
+
+    return 0;
+};
+
+SETTINGS_STATIC_HANDLER_DEFINE(physical_layouts, "physical_layouts", NULL,
+                               physical_layouts_handle_set, NULL, NULL);
+
+#endif // IS_ENABLED(CONFIG_SETTINGS)
+
+static int zmk_physical_layouts_init(void) {
+    k_work_init(&msg_processor.work, zmk_physical_layouts_kscan_process_msgq);
+
+#if IS_ENABLED(CONFIG_PM_DEVICE)
+    for (int l = 0; l < ARRAY_SIZE(layouts); l++) {
+        const struct zmk_physical_layout *pl = layouts[l];
+        if (pl->kscan) {
+            if (pm_device_wakeup_is_capable(pl->kscan)) {
+                pm_device_wakeup_enable(pl->kscan, true);
+            }
+        }
+    }
+#endif // IS_ENABLED(CONFIG_PM_DEVICE)
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+    settings_subsys_init();
```

Comment:
Is this necessary with it being called in `main()` now?

Reply:
Nope. Nuking.